source/extensions/common/tap/tap_config_base.cc

```diff
@@ -38,20 +65,85 @@ Matcher& TapConfigBaseImpl::rootMatcher() {
   return *matchers_[0];
 }
 
+namespace {
+void swapBytesToString(envoy::data::tap::v2alpha::Body& body) {
+  body.set_allocated_as_string(body.release_as_bytes());
+}
+} // namespace
+
+void TapConfigBaseImpl::submitBufferedTrace(
+    const std::shared_ptr<envoy::data::tap::v2alpha::BufferedTraceWrapper>& trace,
+    uint64_t trace_id) {
+  // Swap the "bytes" string into the "string" string. This is done purely so that JSON
+  // serialization will serialize as a string vs. doing base64 encoding.
+  if (sink_format_ == envoy::service::tap::v2alpha::OutputSink::JSON_BODY_AS_STRING) {
+    switch (trace->trace_case()) {
+    case envoy::data::tap::v2alpha::BufferedTraceWrapper::kHttpBufferedTrace: {
+      auto* http_trace = trace->mutable_http_buffered_trace();
+      if (http_trace->has_request() && http_trace->request().has_body()) {
+        swapBytesToString(*http_trace->mutable_request()->mutable_body());
+      }
+      if (http_trace->has_response() && http_trace->response().has_body()) {
+        swapBytesToString(*http_trace->mutable_response()->mutable_body());
+      }
+      break;
+    }
+    case envoy::data::tap::v2alpha::BufferedTraceWrapper::kSocketBufferedTrace: {
+      auto* socket_trace = trace->mutable_socket_buffered_trace();
+      for (auto& event : *socket_trace->mutable_events()) {
+        if (event.has_read()) {
+          swapBytesToString(*event.mutable_read()->mutable_data());
+        } else {
+          ASSERT(event.has_write());
+          swapBytesToString(*event.mutable_write()->mutable_data());
+        }
+      }
+      break;
+    }
+    case envoy::data::tap::v2alpha::BufferedTraceWrapper::TRACE_NOT_SET:
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+  }
+
+  sink_to_use_->submitBufferedTrace(trace, sink_format_, trace_id);
+}
+
 void FilePerTapSink::submitBufferedTrace(
-    std::shared_ptr<envoy::data::tap::v2alpha::BufferedTraceWrapper> trace, uint64_t trace_id) {
-  // TODO(mattklein123): Add JSON format.
-  const bool text_format =
-      config_.format() == envoy::service::tap::v2alpha::FilePerTapSink::PROTO_TEXT;
-  const std::string path =
-      fmt::format("{}_{}.{}", config_.path_prefix(), trace_id, text_format ? "pb_text" : "pb");
+    const std::shared_ptr<envoy::data::tap::v2alpha::BufferedTraceWrapper>& trace,
+    envoy::service::tap::v2alpha::OutputSink::Format format, uint64_t trace_id) {
+  std::string path = fmt::format("{}_{}", config_.path_prefix(), trace_id);
+  switch (format) {
+  case envoy::service::tap::v2alpha::OutputSink::PROTO_BINARY:
+    path += ".pb";
+    break;
+  case envoy::service::tap::v2alpha::OutputSink::PROTO_TEXT:
+    path += ".pb_text";
```

Comment:
These suffixes could be constants that we standardize with our proto ingestion in `MessageUtil`.